[TRTLLM-6611][feat] Add warnings and stricter validation to LoraManager adapter loading

venkywonka opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error handling and validation when loading NeMo and HuggingFace LoRA adapters, with clear error messages for incomplete or corrupted checkpoints.
  * Added warnings for unsupported or skipped LoRA modules during adapter loading.

* **New Features**
  * Enhanced user feedback with warnings and status messages during the LoRA adapter loading process.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

[JIRA ticket/NVBugs ID/GitHub issue][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR to support a new feature about cache manager for JIRA ticket TRTLLM-1000, it would be like:

[TRTLLM-1000][feat] Support a new feature about cache manager

Or I have a PR to fix a Llama3 accuracy issue:

[https://nvbugs/1234567][fix] Fix Llama3 accuracy issue
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

venkywonka committed:
add loggign

Signed-off-by: Venky Ganesh <23023424+venkywonka@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #13418](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13418/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10037](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10037/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #13471](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13471/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10085](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10085/) completed with status: 'FAILURE'

shaharmor98 commented on tensorrt_llm/lora_manager.py:
```diff
@@ -1022,10 +1048,16 @@ def load_from_model_dir(uid, model_dir, hf_config):
-                        is_moe = True
+
+                    # Check if this is MoE by looking for integer keys (expert indices)
+                    has_expert_indices = all(isinstance(k, int) for k in module_weights.keys())
```
I don’t have a suggestion and don’t necessarily think you’re wrong, but I was wondering if this check isn’t risky.
Could there be scenarios where integers don’t belong to MoE?

venkywonka replied:
Honestly i don't know myself - but the prior `is_moe` check was far worse 😅 since the code only seems to support 2 kinds of dict schema anyways (the moe and non-moe kind), i believe this works for now. In the newest commit, i have added stricter validation for what it means to be moe (all keys must be integer, and value must be a nested dicts)

shaharmor98 commented on tensorrt_llm/lora_manager.py:
```diff
@@ -1044,7 +1076,23 @@ def load_from_model_dir(uid, model_dir, hf_config):
+                        if "out" not in module_weights:
+                            missing_matrices.append("'out' matrix (lora_B)")
+
+                        if missing_matrices:
```
Overall I'm really in favor of this change, I think it's much more robust.  
I already commented on something similar before, my main comment here is that we can just create a utility function, which can (and should) be shared between hf and nemo lora loading methods, that asserts the validity of the loaded module.  
Otherwise, I feel there’s a big chunk of duplicated code for no good reason.

venkywonka replied:
agreed!

shaharmor98 reviewed (commented):
Looks really good, thank you for the good work  
Left some small comments

venkywonka committed:
Enhance LoRA manager with validation checks and logging

- Added logging functionality to track loading of NeMo LoRA adapters.
- Introduced validation functions to ensure required 'in' and 'out' matrices are present for LoRA modules.
- Updated handling of supported modules to use a defined set, improving maintainability and clarity.
- Refactored code to reduce redundancy in matrix validation logic.

Signed-off-by: Venky Ganesh <23023424+venkywonka@users.noreply.github.com>

venkywonka commented on tensorrt_llm/lora_manager.py:
```diff
@@ -832,10 +833,33 @@ def load_from_model_file(uid, model_file):
+                        warnings.warn(
+                            f"LoRA module '{lora_module}' not supported in NeMo loading for "
+                            f"layer {layer_idx}, skipping. NeMo LoRA currently only supports "
+                            f"'attn_qkv' modules."
```
ah i have parametrized this so there's single point of point of change. thanks!

venkywonka commented on tensorrt_llm/lora_manager.py:
```diff
@@ -832,10 +833,33 @@ def load_from_model_file(uid, model_file):
+                        # Check for missing "in" matrix (lora_A equivalent)
+                        if "in" not in all_lora_weights[layer_idx]:
+                            raise ValueError(
+                                f"Layer {layer_idx} is missing required 'in' matrix (lora_A equivalent) for attn_qkv "
```
yep agreed, thanks for the catch! i have condensed it to a helper that i have reused in 3 places. thanks!

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13582](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13582/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13582](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13582/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10182](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10182/) completed with status: 'FAILURE'

shaharmor98 reviewed (approved):
LGTM

shaharmor98 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13621](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13621/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10218](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10218/) completed with status: 'FAILURE'

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13718](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13718/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13718](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13718/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10307](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10307/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/10307/test-results/rerun_report.html) for details.
